Trailing comma implementation

mateusrodriguesxyz opened the pull request:
This is a prototype implementation for [Allow trailing comma in tuples, arguments and if/guard/while conditions](https://github.com/apple/swift-evolution/pull/2344) gated behind `-enable-experimental-feature TrailingComma`.

ahoppen commented on Sources/SwiftParser/Declarations.swift:
```diff
@@ -476,6 +476,10 @@ extension Parser {
             arena: self.arena
           )
         )
+        if self.at(prefix: ">") {
+          break
+        }
+            
```
Why is this needed? Could you add a comment for it?

mateusrodriguesxyz replied:
Actually I was trying trailing comma for generics and forget to remove this.

ahoppen commented on Sources/SwiftParser/Expressions.swift:
```diff
@@ -1853,7 +1853,7 @@ extension Parser {
   /// This is currently the same as parsing a tuple expression. In the future,
   /// this will be a dedicated argument list type.
-  mutating func parseArgumentListElements(pattern: PatternContext, flavor: ExprFlavor = .basic) -> [RawLabeledExprSyntax] {
+  mutating func parseArgumentListElements(pattern: PatternContext, flavor: ExprFlavor = .basic, allowTrailingComma: Bool = true) -> [RawLabeledExprSyntax] {
```
I’m generally not a fan of defaulted arguments for these kinds of options because it’s too easy to just forget to pass it down from a different call or not think about it. I’d prefer to be explicit here and require it in the calls.

ahoppen commented on Sources/SwiftParser/Expressions.swift:
```diff
@@ -2088,14 +2093,33 @@ extension Parser {
 
 extension Parser {
+  
+  func extractCodeBlockFromLast(_ conditions: inout RawConditionElementListSyntax) -> RawCodeBlockSyntax? {
+    if let closure = conditions.elements.last?.condition.as(RawClosureExprSyntax.self), closure.signature == nil {
+      var elements = conditions.elements
+      elements.removeLast()
+      conditions = RawConditionElementListSyntax(elements: elements, arena: self.arena)
+      return RawCodeBlockSyntax(
+        closure.unexpectedBeforeLeftBrace,
+        leftBrace: closure.leftBrace,
+        statements: closure.statements,
+        closure.unexpectedBetweenStatementsAndRightBrace,
+        rightBrace: closure.rightBrace,
+        closure.unexpectedAfterRightBrace,
+        arena: self.arena
+      )
+    } else {
+      return nil
+    }
+  }
```
I would prefer if we didn’t have to do this. Because all the syntax nodes are allocated in a `SyntaxArena` and not freed until the arena gets deallocated, the original `RawConditionElementListSyntax` will continue to stick around and use memory. 

We sometimes modify already parsed syntax nodes in case there is a syntax error but in that case we say that using a little memory is OK because source files usually don’t have many errors. But in valid code, I would prefer to avoid it. It probably means that you need to implement some lookahead logic to determine whether an item is a closure or the if body.

ahoppen commented on Sources/SwiftParser/Expressions.swift:
```diff
@@ -2110,8 +2134,22 @@ extension Parser {
+    var body: RawCodeBlockSyntax
+    
+    if experimentalFeatures.contains(.trailingComma) {
+      if !self.at(prefix: "{"), let blockAfterTrailingComma = extractCodeBlockFromLast(&conditions) {
```
Should be able to simplify this slightly.

```suggestion
      if !self.at(.leftBrace), let blockAfterTrailingComma = extractCodeBlockFromLast(&conditions) {
```

ahoppen commented on Sources/SwiftParser/Statements.swift:
```diff
@@ -162,6 +162,11 @@ extension Parser {
     var loopProgress = LoopProgressCondition()
     repeat {
+      if experimentalFeatures.contains(.trailingComma) {
+        if elements.count > 0, currentToken.tokenText == "else" {
```
```suggestion
        if elements.count > 0, self.at(.else) {
```

Also, is the checking for `else` needed for `guard` statements? Would be good to add a comment for that, I think.

mateusrodriguesxyz replied:
That's right, I will add a comment. But maybe `self.at(.keyword(.else)), self.peek(isAt: .leftBrace)` is a better terminator for `guard`?

ahoppen replied:
I think just checking for `else` is sufficient. Just a comment would be good.

ahoppen commented on Sources/SwiftParser/Statements.swift:
```diff
@@ -181,6 +186,44 @@ extension Parser {
   }
+  
+  /// Parse a list of condition elements returning optional trailing comma block.
+  mutating func _parseConditionListWithTrailingCommaBlock() -> (conditions: RawConditionElementListSyntax, blockAfterTrailingComma: RawClosureExprSyntax?) {
```
I don’t think this is being used anywhere, is it?

mateusrodriguesxyz replied:
You're right,

ahoppen commented on Tests/SwiftParserTest/TrailingCommaTests.swift:
```diff
@@ -0,0 +1,96 @@
+    )
+    assertParse("while conditionA, {x}(), { value += 1 }", experimentalFeatures: .trailingComma)
+  }
+  
```
Another interesting test case is

```swift
if foo { print("a") }, else: { print("b") } {
}
```

Where `else:` is supposed to be an additional trailing closure, `foo` being defined as follows

```swift
func foo(_: () -> Void, `else`: () -> Void) -> Bool { true }
```

Also worth adding a test case where the trailing closure label is something normal like `bar`.

mateusrodriguesxyz replied:
By the way, currently `if foo { print("a") } else: { print("b") } { }` doesn't parse correctly. Is this intended or should be fixed by modifying [atValidTrailingClosure](https://github.com/apple/swift-syntax/blob/25ce3a297818076ceab8054873a6189e8b8b6688/Sources/SwiftParser/Expressions.swift#L1983) when flavor is `stmtCondition`?

ahoppen replied:
Define “correctly” ;-). It’s a super weird edge case and I think the current behavior or parsing `else` as the start of an `else` block is the best solution IMO. Parsing `else` as an additional trailing closure makes thinks even more confusing IMO. 

I would mostly like to make sure that we don’t change the current behavior.

ahoppen reviewed (commented):
Some initial thoughts. My biggest comment is that we should implement the disambiguation between an additional condition element and the body using lookahead instead of by modifying the tree.

mateusrodriguesxyz commented:
@ahoppen I've updated the implementation for conditionals to use a much simpler lookahead approach

ahoppen commented on Sources/SwiftParser/Statements.swift:
```diff
@@ -1038,4 +1066,39 @@ extension Parser.Lookahead {
+        if !self.atStartOfLine, self.at(.leftParen) || self.at(.leftBrace) {
+          return false
+        }
+        return true
```
Sorry that I’m in the “Let me go and break your code mood. The following is currently parsed as an expression but would now be parsed as a body

```swift
if true, { (x: () -> Void) in return true} != nil {
}
```

But I think this might be sufficiently stupid that we can ignore it.

CC @bnbarham @hamishknight @rintaro Would like to hear your opinions here as well.

mateusrodriguesxyz replied:
I'm experimenting with changing `atStartOfStatementBody()` to check for `true` cases instead. I think correct bodies are less error prone so maybe there's less risk of changing existing code.

```swift
mutating func atStartOfStatementBody() -> Bool {
  guard at(.leftBrace) else {
    return false
  }
  skipSingle()
  if self.at(.endOfFile) || self.at(.semicolon, .keyword(.else)) || self.at(.rightBrace, .rightParen) {
    return true
  }
  if self.atStartOfLine {
    return !withLookahead { $0.atStartOfStatementBody() } && !self.at(.binaryOperator)
  }
  return false
}
```

The following statements are all parsing the intended body:

```swift
if true, { print("body") }

if true, { print("body") };

if true, { print("body") } else { }

{ if true, { print("body") } }

( if true, { print("body") } )

if true, { true }, { print("body") }

if true, { true } { print("body") }

if true, { true }
{ print("body") }

if true, { (x: () -> Void) in return true} != nil { print("body") }

if true, { (x: () -> Void) in return true }
!= nil { print("body") }

```

What do you think? @ahoppen

ahoppen replied:
Yes, I think this is a lot safer and easier to follow.

I just went through the code, added some explanations (please verify that they are correct) and some questions as well

```swift
mutating func atStartOfStatementBody() -> Bool {
  guard at(.leftBrace) else {
    // Statement bodies always start with a '{'. If there is no '{', we can't be at the statement body
    return false
  }
  skipSingle()
  if self.at(.endOfFile) {
    // There's nothing else in the source file, which could be the statement body. So this must be it.
  }
  if self.at(.semicolon) {
    // We can't have a semicolon between the condition and the statement body. So this must be the statement body.
    return true
  }
  if self.at(.keyword(.else)) {
    // If the next token is an `else` keyword, this must be the statement body of an `if` statement since the condition cannot be followed by `else`.
    // Note that `guard` condition does not u se `atStartOfStatementBody`
    return true
  }
    
  if self.at(.rightBrace) {
    // What is this needed for? Do you have an explanation?
    return true
  }
  if self.at(.rightParen) {
    // What is this condition for? Do you have an explanation?
    return true
  }
  if self.atStartOfLine {
    // If this is followed by a newline, it is most likely a statement body. The only exceptions are
    if withLookahead({ $0.atStartOfStatementBody() }) {
      // If a statement body starts after this one. In this case we have something like
      // if takesTrailing { print("trailing closure") }
      // {
      //   print("if body")
      // }
      return false
    }
    if self.at(.binaryOperator) {
      // If we have a binary operator. An if/switch expression cannot be the left-hand side of an operator, so this cannot be the if body, eg.
      // if takesTrailing { print("trailing closure") }
      // != nil 
      // {
      //   print("body")
      // }
      return false
    }
    // Does this correctly handle something like the following?
    // if takesTrailing { print("trailing closure") }
    //   , true {
    //  
    // }

    // I think it might be safer if we could also somehow flip this into a whitelist instead of blacklist rule.
    // Ie. only return `true` if the next token is at the start of a line *and* satisfies some condition.
    return true
  }
  return false
}
```

mateusrodriguesxyz replied:
The comments are great! It’s okay to add them to the PR?

I use `at(.rightBrace)` for a case like `if true,  {  if true,  { }  }` and `at(. rightParen)` for if statements surrounded by parenthesis (It’s not valid code but without handling this there may be regression in diagnostics). 

I’m pretty sure we are safe for trailing closure because the function name will prevent the lookahead from going beyond `guard at(.leftBrace)`.

By “flip this into a whitelist” do you mean came up with a condition that returns true to assert that it’s the body? There are so many possibilities for the token at the start of a line that I can only think “is at the start of a line *and* does **not** satisfies some condition.”

mateusrodriguesxyz replied:
I'm a little confused 

> Which of my comments does this refer to?

I was refering to
```swift
if takesTrailing { print("trailing closure") }
, true {
 
}
```

I'm confused with your examples using trailing closure because in my tests they don't go past beyond  `guard at(.leftBrace)` because the function name, only closures like `if true, { print("closure") } { print("body") }` will.

The `.doStuff` is something returning a `Bool` right? Your example doesn't even compile today which took me by surprise 😮 It compiles though if the `.doStuff` is on the same line.

```swift
if takesTrailing { print("trailing") }
    .doStuff {
}
// ❌ Cannot convert value of type '(() -> Void) -> Foo' to expected condition type 'Bool'
// ❌ Reference to member 'doStuff' cannot be resolved without a contextual type

if takesTrailing { print("trailing") }.doStuff { // ✅
}
```

ahoppen replied:
> I'm confused with your examples using trailing closure because in my tests they don't go past beyond `guard at(.leftBrace)` because the function name, only closures like `if true, { print("closure") } { print("body") }` will.

Oh, I think my brain just switched modes and I started thinking about trailing closure disambiguation instead of first-level closures. Sorry about that. Disregard most things I said. The only meaningful examples that I have would be 

```swift
if true,
nil != { return true }
, true {
}
```

And silly postfix operators, which luckily need to be on the same line. But maybe good to add a test case for this as well.

```swift
postfix operator +++

postfix func +++(_ x: () -> Bool) -> Bool { true }

if true,
{ return true }+++ {
}
```

ahoppen commented on Tests/SwiftParserTest/TrailingCommaTests.swift:
```diff
@@ -0,0 +1,130 @@
+import XCTest
+
+final class TrailingCommaTests: ParserTestCase {
+  
```
Superfluous newline

ahoppen commented on Tests/SwiftParserTest/TrailingCommaTests.swift:
```diff
@@ -0,0 +1,130 @@
+final class TrailingCommaTests: ParserTestCase {
+  
+  func testTuple() {
+    
```
Superfluous newline (same in the other test methods)

ahoppen reviewed (commented):
Thank you. The approach is the right one, I think. Getting the lookahead logic right is the key building block here, I think and we are on a very good track. I needed to bang my head really hard to come up with counterexamples, those are very contrived and maybe we can ignore them.

mateusrodriguesxyz commented on Sources/SwiftParser/Expressions.swift:
```diff
@@ -2033,6 +2048,10 @@ extension Parser.Lookahead {
     guard lookahead.consume(if: .rightBrace) != nil else {
       return false
     }
+    
+    if lookahead.atStartOfLabelledTrailingClosure() {
+      return true
+    }
```
This is a change that should not have been added. I was experimenting with lifting the restriction of multiple trailing closures in conditions. I will remove it.

mateusrodriguesxyz commented on Sources/SwiftParser/Statements.swift:
```diff
@@ -154,21 +154,42 @@ extension Parser {
+        elements.count == 1,
+        condition.is(RawMissingExprSyntax.self)
+      {
+        continue
```
Indeed `break` is better. I've added this code so  `guard true,`  would not report `expected expression in 'guard' statement`.

mateusrodriguesxyz commented on Sources/SwiftParser/Expressions.swift:
```diff
@@ -1875,6 +1875,11 @@ extension Parser {
     var keepGoing: RawTokenSyntax? = nil
     var loopProgress = LoopProgressCondition()
     repeat {
+      if experimentalFeatures.contains(.trailingComma) {
+        if allowTrailingComma, currentToken.rawTokenKind == .rightParen {
+          break
+        }
+      }
```
Do you mean something like this?

```swift
mutating func atConditionListTerminator(isGuardStatement: Bool) -> Bool {
    if experimentalFeatures.contains(.trailingComma) {
      return isGuardStatement ? self.at(.keyword(.else)) : withLookahead({ $0.atStartOfStatementBody() })
    } else {
      return false
    }
  }
```

```swift
while keepGoing != nil && self.hasProgressed(&loopProgress) && !atConditionListTerminator(isGuardStatement: isGuardStatement)
```

ahoppen replied:
Yeah, I think something like this is good. I’d just stick `hasProgressed` to the end of the condition.

mateusrodriguesxyz commented:
@ahoppen I've fixed minor problems with parseArgumentListElements and updated parseConditionList and atStartOfStatementBody implementations to address our conversation. I've also updated the tests and included some substructure assertions to ensure that the expected block is parsed as the statement body.

ahoppen commented on Sources/SwiftParser/Expressions.swift:
```diff
@@ -1875,6 +1879,9 @@ extension Parser {
     var keepGoing: RawTokenSyntax? = nil
     var loopProgress = LoopProgressCondition()
     repeat {
+      if allowTrailingComma, currentToken.rawTokenKind == .rightParen {
```
Similar comment as in other places, I would prefer this to be centralized in the `repeat while` condition.

Also

```suggestion
      if allowTrailingComma, self.at(.rightParen) {
```

ahoppen commented on Sources/SwiftParser/Expressions.swift:
```diff
@@ -2089,9 +2106,8 @@ extension Parser {
-    ifHandle: RecoveryConsumptionHandle
-  ) -> RawIfExprSyntax {
+  mutating func parseIfExpression(ifHandle: RecoveryConsumptionHandle) -> RawIfExprSyntax {
+
```
Superfluous newline

ahoppen reviewed (commented):
I think we’re on the final stretch for this PR. Thanks for your continued work on this.

mateusrodriguesxyz commented on Tests/SwiftParserTest/TrailingCommaTests.swift:
```diff
@@ -0,0 +1,271 @@
+import XCTest
+
+final class TrailingCommaTests: ParserTestCase {
+  func testTuple() {
```
That's cool, much better!

ahoppen commented on Tests/SwiftParserTest/TrailingCommaTests.swift:
```diff
@@ -15,103 +15,62 @@ import SwiftSyntax
         conditions: [
-          ConditionElementSyntax.init(condition: .init(ExprSyntax(stringLiteral: "true")), trailingComma: .commaToken()),
-          ConditionElementSyntax.init(condition: .init(ExprSyntax(stringLiteral: "{ true }()")), trailingComma: nil),
+            ConditionElementSyntax(condition: .expression("true"), trailingComma: .commaToken()),
```
Looks like this has some superfluous indentation. Could you run `./swift-syntax-dev-utils format`?

ahoppen reviewed (approved):
Nice. Looking good to me 👍🏽 

I would still like to hear a second opinion from @rintaro on the implementation.

mateusrodriguesxyz commented on Sources/SwiftParser/Attributes.swift:
```diff
@@ -365,7 +365,7 @@ extension Parser {
     case nil:
       return parseAttribute(argumentMode: .customAttribute) { parser in
-        let arguments = parser.parseArgumentListElements(pattern: .none)
+        let arguments = parser.parseArgumentListElements(pattern: .none, allowTrailingComma: false)
```
So I'm revisiting **argumentListElements** usages and those are all that I've tracked: subscripts, attributes, macro roles, macro expansions and string literals.

Do you see any problems regardless of trailing comma in **string literals**?

Besides these there's one more that I can't figure out in [parseKeyPathExpression](https://github.com/apple/swift-syntax/blob/ec4e75afe680ac43540ce5412ea8bc294543c513/Sources/SwiftParser/Expressions.swift#L1013).

ahoppen replied:
I don’t see an inherent problem with allowing trailing commas in string interpolation. Could we maybe keep it as a separate, follow-up PR though? We’ve iterated on this one quite a bit already and I think it would be easier if we could focus on landing this one first and then consider the implications of expanding trailing closures to other areas in a separate PR.

And the following should be an example of a key path subscript that contains a comma:

```swift
struct Foo {
  subscript(x: Int, y y: Int) -> Int { 1 }
}

func test() {
  let _: KeyPath<Foo, Int> = \.[1, y: 2]
}
```

mateusrodriguesxyz replied:
So we leave the PR as it is?

ahoppen replied:
Yes, that would be my suggestion.

rintaro commented:
Could you let me know the plan for other comma separated lists like func parameter list, tuple type, generic paremeter/argument list etc?
I believe your proposal includes parameter list support at least, right?

ahoppen commented:
I just suggested to keep the PR as-is for now and implement trailing closures for other constructs in a follow-up PR (https://github.com/apple/swift-syntax/pull/2515#discussion_r1530160204). I’ve already spent considerable time reviewing this PR and it’s easier review-wise if we can consider the implications of trailing commas for other constructs in a separate PR.

mateusrodriguesxyz commented:
> I believe your proposal includes parameter list support at least, right?

That's right. The current proposal and implementation includes tuples, functions parameter/argument list and `if`, `guard` and `while` conditions. I'm keeping minimal and included the ones that I consider the most demanded use cases.

mateusrodriguesxyz commented:
BTW @ahoppen thank you so much for all the help and patience while reviewing this PR and sorry for all the troubles. That's my first time trying to contribute to the language and whatever the outcome I'm happy with all that I've learned.

rintaro commented on Sources/SwiftParser/Statements.swift:
```diff
@@ -1038,4 +1051,63 @@ extension Parser.Lookahead {
+      return true
+    }
+    if self.atStartOfLine {
+      // If the current token is at the start of a line, it is most likely a statement body. The only exceptions are:
+      if withLookahead({ $0.atStartOfStatementBody() }) {
+        // If a statement body starts after this on, we actually have a closure followed by the statement body e.g.
+        // if true, { true }
+        // {
+        //   print("body")
+        // }
+        return false
```
Sorry if I missed the previous discussion, but do we really care this case? Closure alone cannot be a condition. it must be used/called some way (e.g. call it, or make it a part of a sequence expression). This example case is ambiguous, it can be missing `()` after `{ true }` or missing `()` after `{ print("body") }`. I don't think it's worth to lookahead again if there's any strong reason.

mateusrodriguesxyz replied:
I agree that's a odd case  but today that's a valid and the parser reports a `Function produces expected type 'Bool'; did you mean to call it with '()'?` error so without this lookahead there would be a regression.

rintaro replied:
That's not a valid condition today, it's just parsed as a condition expression and diagnosed in type checker. Since it's invalid anyway, I wouldn't consider it's a regression even if we don't parse it as a closure.

mateusrodriguesxyz replied:
I don't feel comfortable changing existing behavior and as a user I consider the type checker diagnostic more helpful because is probably more close to the user intention.

mateusrodriguesxyz replied:
I can add `at(.leftBrace)` before the `withLookahead` to avoid actually looking ahead if not necessary.

rintaro replied:
But that's _before_ the trailing comma support, so there was no ambiguity. Supporting trailing comma introduces some ambiguity, and for example, I believe the user intension is that `{ return true }` is the body in the following case:
```swift
func test() -> Bool {
  if
    condition1,
    condition2,
  {
    return true
  }
  
  { print("something") }
}
```

mateusrodriguesxyz replied:
Sorry, I took I time from this PR and missed your last answer. That's a really good counterexample. Honestly, I'm good with either behavior, whatever is less controversy. Could you weight in @ahoppen?

ahoppen replied:
I agree with @rintaro.

mateusrodriguesxyz replied:
I will change then and also resolve the conflicts with main branch.

rintaro commented:
> The current proposal and implementation includes tuples, functions parameter/argument list and if, guard and while conditions.

I don't see implementation/tests for functions parameter. e.g. from your proposal
```
func foo(
    a: Int = 0, 
    b: Int = 0, 
) {
}
```
Am I missing anything?

rintaro commented:
To be clear I'm not arguing that this PR should include parameter list support. I just want to make sure I'm not missing anything, and know the plan.

mateusrodriguesxyz commented:
> I don't see implementation/tests for functions _parameter_ list. e.g. from your proposal

I think I didn't add this test case because arguments and parameters are parsed by the same `parsePostfixExpressionSuffix` so testing arguments covers parameters, but I can add it to make it clear.

mateusrodriguesxyz commented on Sources/SwiftParser/Statements.swift:
```diff
@@ -1038,4 +1051,63 @@ extension Parser.Lookahead {
   }
+
+  /// Returns `true` if the current token represents the start of an `if` or `while` statement body.
+  mutating func atStartOfStatementBody() -> Bool {
```
You're right. I'm bad with naming, your suggestion is more clear. I just tried replacing it by `atValidTrailingClosure` and most of my test cases failed. I don't think it worth trying to consolidating them, at least not now, because it's seems there are a few differences and I don't feel comfortable changing a function that's used in places outside the scope of my PR.

rintaro replied:
Okay

rintaro commented:
> arguments and parameters are parsed by the same parsePostfixExpressionSuffix 

Parameters are parsed with `parseParameterClause()` not `parsePostfixExpressionSuffix()` :)

rintaro commented on Sources/SwiftParser/Statements.swift:
```diff
@@ -177,11 +178,22 @@ extension Parser {
+    guard experimentalFeatures.contains(.trailingComma) else {
+      return false
+    }
+    // condition terminator is `else` for `guard` statements and start of statement body for `if` or `while` statements
+    if isGuardStatement {
+      return self.at(.keyword(.else))
+    }
```
Missing `else` is a very common mistakes for `guard` statement. Probably parsing it as the body emit better diagnostics if it's looks like a body.
```swift
    if isGuardStatement {
      if self.at(.keyword(.else)) {
        return true
      }
      // Missing `else` is a common mistake, fall back to  lookahead `{` check.
    }
```

mateusrodriguesxyz commented:
> Parameters are parsed with `parseParameterClause()` not `parsePostfixExpressionSuffix()` :)

Now that's strange because I never touched this functions but if I add your example as a test case `parsePostfixExpressionSuffix` is called and it does parse correctly so I assumed that was it 🤔

rintaro commented:
That's... concerning (for current parser) 😅

unnamedd commented on Sources/SwiftParser/Expressions.swift:
```diff
@@ -1911,9 +1915,17 @@ extension Parser {
   }
+
+  mutating func atArgumentListTerminator(_ allowTrailingComma: Bool) -> Bool {
+    if allowTrailingComma, self.at(.rightParen) {
+      return true
+    } else {
+      return false
+    }
```
I was wondering if wouldn't be better if, instead of writing if/else to return a boolean, if wouldn't be better to write just one line of code, like this:
```suggestion
    allowTrailingComma && self.at(.rightParen)
```

mateusrodriguesxyz commented:
> That's... concerning (for current parser) 😅

I think `parseParameterClause`already supports trailing comma because `!self.at(.endOfFile, .rightParen)` breaks the while even if the last parameter has as trailing comma so I think ` `parsePostfixExpressionSuffix` was the one blocking.

https://github.com/apple/swift-syntax/blob/18cd7f90af1615eec95e173d84823a970b7469c6/Sources/SwiftParser/Parameters.swift#L291-L302

rintaro commented:
Yeah, I mean we've just discovered a bug in the current parser. We _should_ disallow it  when `trailingComma` experimental feature is not enabled.
(I don't think it's a scope of this PR)

mateusrodriguesxyz commented:
I've update https://github.com/apple/swift/pull/71975 implementation to match this PR.

mateusrodriguesxyz commented:
I've update `atStartOfStatementBody` implementation according and resolved the conflicts. I think that's it? @ahoppen @rintaro Thank you both for the review!

ahoppen commented:
One last thing: Could you squash your commits? It makes for a nicer Git history https://github.com/apple/swift-syntax/blob/main/CONTRIBUTING.md#authoring-commits

Other than that, there’s anything to be done from my side. I would like to wait for @rintaro’s approval as well though.

mateusrodriguesxyz commented:
@rintaro do you have any more to add before I squash the commits?

rintaro commented:
@swift-ci Please test

mateusrodriguesxyz commented:
@rintaro I had forget to format the source code after I merged with main, sorry. I think it will pass all the checks now.

rintaro commented:
@swift-ci Please test

ahoppen commented:
@swift-ci Please test Windows

mateusrodriguesxyz commented:
> One last thing: Could you squash your commits?

@ahoppen done!

kimdv commented:
@swift-ci please test

kimdv commented:
@swift-ci please test windows
